main.go

```diff
@@ -238,74 +318,124 @@ func main() {
+			log.WithField("address", address).
+				Info("listening on address")
 
-		if strings.Index(listenAddr, "://") == -1 {
-			log.Printf("Listen on %s ...\n", listenAddr)
-
-			lsnr, err = net.Listen("tcp", listenAddr)
-		} else if strings.HasPrefix(listenAddr, "starttls://") {
-			listenAddr = strings.TrimPrefix(listenAddr, "starttls://")
+			listener := listenOnAddr(server, address)
+			listeners = append(listeners, listener)
+		} else if strings.HasPrefix(addresses[i], "starttls://") {
+			address = strings.TrimPrefix(address, "starttls://")
 
 			server.TLSConfig = getTLSConfig()
 			server.ForceTLS = *localForceTLS
 
-			log.Printf("Listen on %s (STARTTLS) ...\n", listenAddr)
-			lsnr, err = net.Listen("tcp", listenAddr)
-		} else if strings.HasPrefix(listenAddr, "tls://") {
-			listenAddr = strings.TrimPrefix(listenAddr, "tls://")
+			log.WithField("address", address).
+				Info("listening on STARTTLS address")
+
+			listener := listenOnAddr(server, address)
+			listeners = append(listeners, listener)
+		} else if strings.HasPrefix(addresses[i], "tls://") {
 
+			address = strings.TrimPrefix(address, "tls://")
 			server.TLSConfig = getTLSConfig()
 
-			log.Printf("Listen on %s (TLS) ...\n", listenAddr)
-			lsnr, err = tls.Listen("tcp", listenAddr, server.TLSConfig)
+			log.WithField("address", address).
+				Info("listening on TLS address")
+
+			listener := listenOnTLSAddr(server, address, server.TLSConfig)
+			listeners = append(listeners, listener)
+
 		} else {
-			log.Fatal("Unknown protocol in listen address ", listenAddr)
+			log.WithField("address", address).
+				Fatal("unknown protocol in address")
 		}
+	}
 
-		if err != nil {
-			log.Fatal(err)
-		}
-		defer lsnr.Close()
+	handleSignals(listeners)
+}
 
-		go server.Serve(lsnr)
+func listenOnAddr(server *smtpd.Server, addr string) net.Listener {
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.WithField("err", err).
+			WithField("address", addr).
+			Fatal("could not listen on address")
 	}
 
-	for true {
-		time.Sleep(time.Minute)
+	go server.Serve(listener)
+	return listener
+}
+
+func listenOnTLSAddr(server *smtpd.Server, addr string, config *tls.Config) net.Listener {
+	listener, err := tls.Listen("tcp", addr, config)
+	if err != nil {
+		log.WithField("err", err).
+			WithField("address", addr).
+			Fatal("could not listen on address")
 	}
+
+	go server.Serve(listener)
+	return listener
+}
+
+func handleSignals(servers []net.Listener) {
+	sigs := make(chan os.Signal, 1)
+	done := make(chan bool, 1)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		sig := <-sigs
+
+		for _, s := range servers {
+			log.WithField("signal", sig).
+				WithField("addr", s.Addr().String()).
+				Warn("closing listener in response to received signal")
+
+			err := s.Close()
+			if err != nil {
+				log.WithField("err", err).
+					Warn("could not close listener")
+			}
+		}
+
+		done <- true
+	}()
+
+	<-done
+	os.Exit(0)
 }
```

Comment:
Perhaps I'm missing something, but I don't see the point of the goroutine or the `done` channel here. The pattern of starting a goroutine just to immediately wait on a channel only sent at the end of that goroutine seems to accomplish nothing. Why not just wait for a signal, shut everything down, and then exit?

```suggestion
func handleSignals(servers []net.Listener) {
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	sig := <-sigs
	
	for _, s := range servers {
		log.WithField("signal", sig).
			WithField("addr", s.Addr().String()).
			Warn("closing listener in response to received signal")
		err := s.Close()
		if err != nil {
			log.WithField("err", err).
				Warn("could not close listener")
		}
	}

	os.Exit(0)
}
```